### DISCUSSION TITLE
Instructions for building locally

### DISCUSSION THREAD

**Discussion by notpeter** [LuaLS/vscode-lua#125] [General]
Hi,

I've spent a couple hours to try and get this extension building locally, but with limited success.  If someone had a basic set of instructions I would happily build-out some proper documentation for this.

In particular use of nested git submodules add significant complexity:

Currently the project has 15 submodules spread across  nested `.gitmodules`:
<pre>
vscode-lua/
├─ <a href="https://github.com/LuaLS/vscode-lua/blob/master/.gitmodules">.gitmodules</a>
|  ├─ <a href="https://github.com/LuaLS/vscode-lua-doc/blob/master/.gitmodules">client/3rd/vscode-lua-doc/.gitmodules</a>
|  ├─ <a href="https://github.com/LuaLS/lua-language-server/blob/master/.gitmodules">server/.gitmodules</a>
|  |  ├─ <a href="https://github.com/actboy168/json.lua/blob/master/.gitmodules">server/3rd/json.lua/.gitmodules</a>
|  |  ├─ <a href="https://github.com/actboy168/luamake/blob/master/.gitmodules">server/3rd/luamake/.gitmodules</a>
</pre>

Resulting in the following directory structure of vscode-lua git dependencies:
<pre>
vscode-lua
├─ client
|  ├─ 3rd
|  |  ├─ vscode-lua-doc*            <a href="https://github.com/LuaLS/vscode-lua-doc">LuaLS/vscode-lua-doc</a>
|  |  |  ├─ doc
|  |  |  |  ├─ zh-cn
|  |  |  |  |  ├─ 53                <a href="https://github.com/cloudwu/lua53doc">cloudwu/lua53doc</a>
|  |  |  ├─ build
|  |  |  |  ├─ 53                   <a href="https://github.com/lua/lua">lua/lua</a>
|  |  |  |  ├─ 54                   <a href="https://github.com/lua/lua">lua/lua</a>
|  ├─ webvue                        <a href="https://github.com/LuaLS/vscode-lua-webvue">LuaLS/vscode-lua-webvue</a>
├─ server*                          <a href="https://github.com/LuaLS/lua-language-server">LuaLS/lua-language-server</a>
|  ├─ 3rd
|  |  ├─ EmmyLuaCodeStyle           <a href="https://github.com/CppCXY/EmmyLuaCodeStyle">CppCXY/EmmyLuaCodeStyle</a>
|  |  ├─ love-api                   <a href="https://github.com/love2d-community/love-api">love2d-community/love-api</a>
|  |  ├─ lovr-api                   <a href="https://github.com/bjornbytes/lovr-docs">bjornbytes/lovr-docs</a>
|  |  ├─ lpeglabel                  <a href="https://github.com/sqmedeiros/lpeglabel">sqmedeiros/lpeglabel</a>
|  |  ├─ luamake*                   <a href="https://github.com/actboy168/luamake">actboy168/luamake</a>
|  |  |  ├─ 3rd
|  |  |  |  ├─ bee.lua              <a href="https://github.com/actboy168/bee.lua">actboy168/bee.lua</a>
|  |  ├─ bee.lua                    <a href="https://github.com/actboy168/bee.lua">actboy168/bee.lua</a>
|  |  ├─ json.lua*                  <a href="https://github.com/actboy168/json.lua">actboy168/json.lua</a>
|  |  |  ├─ test
|  |  |  |  ├─ ltest                <a href="https://github.com/actboy168/ltest">actboy168/ltest</a>
</pre>

I greatly appreciate the hard work of the maintainers (@actboy168, @sumneko, @carsakiller, @CppCXY, @fesily and others) building the lua-language-server and vscode-liua bits. If any of y'all have any rough notes about your setup I would love to write up some documentation to make it possible for others to contribute.  

As is, it's largely inaccessible.  And I say that as someone with pre-existing Lua and VSCode Extension experience.

**Comment by carsakiller** [comment]
Instructions for development and building is something I have wanted to improve for a long time, but, to be honest, I'm not really sure how to get *fully* set up. I don't really know how sumneko does it 😆.

I (think I) know the steps getting set up for developing the VS Code client extension:

You will need `Node.js` and `npm`.

1. Clone `vscode-lua`
2. `cd` into `client/`
3. `git submodule init` to get all submodules
4. `npm i` to install dependencies

You should now have `webvue` (the UI for the addon manager), all the dependencies, and all the code for VS Code extension.

Running `buildClient.sh`/`buildClient.bat` will build the VS Code extension by compiling the TypeScript and building the addon manager UI.

You can use the test task in VS Code to launch an extension debug host for testing changes to the extension.

As far as the server goes, it is a mystery to me. The only help I can offer here is the [build instructions currently in the wiki](https://luals.github.io/wiki/build/), but those are from before I was involved in this project, so not sure if they are helpful.

**Reply by notpeter** [reply]
So I spent a few hours on this today and am still very much in the weeds.

Components:

* [actyboy168/bee.lua](https://github.com/actboy168/bee.lua) - Patched Lua runtime which includes enhancements for debugging and a couple built in libs.
* [actyboy168/luamake](https://github.com/actboy168/luamake) - "make"-like tool written in Lua.
* [ninja-build/ninja](https://github.com/ninja-build/ninja) - Build script used to build luamake.
* [actyboy168/lua-debug](https://github.com/actboy168/lua-debug) - VSCode Extension for Running/Debugging Lua.
* [LuaLS/vscode-lua-doc](https://github.com/LuaLS/vscode-lua-doc) - Generates documentation from Lua53/Lua54 sources. Not a standalone vscode extension, but included in vscode-lua.

luamake requires parts of bee.lua to build itself. 
bee.lua needs a working luamake to build itself. 
There is [no documentation](https://github.com/actboy168/bee.lua/issues/31) for bee.lua.

It's become apparent that Windows is the primary development platform. There are [github actions](https://github.com/LuaLS/vscode-lua/blob/master/.github/workflows/build.yml) creating cross-platform builds, but since the core contributors develop using Windows, the other platforms are less tested.

I was able to post some PRs though 🤷 :
* https://github.com/LuaLS/vscode-lua/pull/126
* https://github.com/actboy168/lua-debug/pull/273
